vcf/io.go

```diff
@@ -54,97 +54,117 @@ func GoReadToChan(filename string) (<-chan Vcf, Header) {
-	//case len(data) == 1:
-	//these lines are sequences, and we are not recording them
-	//case len(line) == 0:
-	//blank line
-	if len(data) < 9 {
-		log.Fatalf("Error when reading this vcf line:\n%s\nExpecting at least 9 columns", line)
-	}
-	curr = Vcf{Chr: data[0], Pos: common.StringToInt(data[1]), Id: data[2], Ref: data[3], Alt: strings.Split(data[4], ","), Filter: data[6], Info: data[7], Format: strings.Split(data[8], ":")}
-	if strings.Compare(data[5], ".") == 0 {
-		curr.Qual = 255
-	} else {
-		curr.Qual = common.StringToFloat64(data[5])
+	var err error
+	data := strings.Split(line, "\t")
+	if len(data) < 8 {
+		log.Fatalf("Error when reading this vcf line:\n%s\nExpecting at least 8 columns", line)
+	}
+
+	curr.Chr = data[0]
+	curr.Pos, err = strconv.Atoi(data[1])
+	if err != nil {
+		log.Fatalf("ERROR: VCF reading\nCould not convert '%s' to an integer in the following line\n%s\n", data[1], line)
+	}
+	curr.Id = data[2]
+	curr.Ref = data[3]
+	curr.Alt = strings.Split(data[4], ",")
+	curr.Qual = 255
+	if data[5] != "." {
+		curr.Qual, err = strconv.ParseFloat(data[5], 64)
+		if err != nil {
+			log.Fatalf("ERROR: VCF reading\nCould not convert '%s' to a float in the following line\n%s\n", data[5], line)
+		}
 	}
-	if len(data) > 9 {
-		//DEBUG: fmt.Println(line)
-		curr.Samples = ParseNotes(data[9], curr.Format)
+	curr.Filter = data[6]
+	curr.Info = data[7]
+
+	if len(data) < 9 {
+		// return if no format field found
+		return curr
 	}
+	curr.Format = strings.Split(data[8], ":")
+	curr.Samples = parseSamples(data[9:], curr.Format, line)
 	return curr
 }
 
-// ParseNotes is a helper function of processVcfLine. Generates a slice of GenomeSample structs from a VCF data line.
-func ParseNotes(data string, format []string) []GenomeSample {
+// parseSamples is a helper function of processVcfLine. Generates a slice of Sample structs from a VCF data line.
+func parseSamples(samples []string, format []string, line string) []Sample {
 	//DEBUG: fmt.Printf("Format: %s. Format[0]: %s.\n", format, format[0])
-	if len(format) == 0 {
-		log.Fatalf("Parsing error: cannot parse notes without formatting information.")
-	}
-	//firstFormat := format[0]
-	//fmt.Println(firstFormat)
-	if format[0] != "GT" && format[0] != "." { //len(format) == 0 checks for when there is info in the notes column but no format specification
-		log.Fatalf("VCF format files with sample information must begin with \"GT\" as the first format column or be marked blank with a period. Here was the first format entry: %s.\nError parsing the line with this Notes column: %s.\n", format[0], data)
+	if format[0] == "." || len(format) == 0 {
+		return nil
 	}
 
-	if format[0] == "." { //if the format column is blank, we do not need to parse further.
-		var blankAnswer []GenomeSample = make([]GenomeSample, 0)
-		return blankAnswer
+	answer := make([]Sample, len(samples))
+	for i := range samples {
+		answer[i].FormatData = strings.Split(samples[i], ":")
+		if format[0] == "GT" {
```

Comment:
so we mandate GT is in 0, but we can still read a VCF that has something other than GT in format[0]? We might want to throw an error in that case